### PR TITLE
Use anyhow crate for error handling

### PR DESCRIPTION
**Pull request by antiphoton**
- In the return value of all functions, change `std::result::Result<T, String>` to `anyhow::Result<T>` .
- In `map_err` closures, use `anyhow::Result::context()` to create error chain.

### CODE REVIEW

File: sxg_rs/src/fetcher/js_fetcher.rs [hunk 1]
```diff
@@ -38,13 +39,13 @@ impl JsFetcher {
...
-        let request = JsValue::from_serde(&request).unwrap();
+    async fn fetch(&self, request: HttpRequest) -> Result<HttpResponse> {
+        let request = JsValue::from_serde(&request).map_err(|e| Error::new(e).context("Failed to parse request."))?;
         let this = JsValue::null();
-        let response = self.0.call1(&this, &request).unwrap();
+        let response = self.0.call1(&this, &request).map_err(|_| Error::msg("JavaScript fetcher throws an error."))?;
         let response = wasm_bindgen_futures::JsFuture::from(js_sys::Promise::from(response));
-        let response = response.await.unwrap();
-        let response: HttpResponse = response.into_serde().unwrap();
+        let response = response.await.map_err(|_| Error::msg("JavaScript fetcher throws an error asynchronously."))?;
```
Comment (banaag) [inline sxg_rs/src/fetcher/js_fetcher.rs:47]: Is it possible to include the error code or detail that was thrown?

File: sxg_rs/src/fetcher/js_fetcher.rs [hunk 2]
```diff
@@ -38,13 +39,13 @@ impl JsFetcher {
 
 #[async_trait(?Send)]
 impl Fetcher for JsFetcher {
-    async fn fetch(&self, request: HttpRequest) -> Result<HttpResponse, String> {
-        let request = JsValue::from_serde(&request).unwrap();
+    async fn fetch(&self, request: HttpRequest) -> Result<HttpResponse> {
+        let request = JsValue::from_serde(&request).map_err(|e| Error::new(e).context("Failed to parse request."))?;
         let this = JsValue::null();
-        let response = self.0.call1(&this, &request).unwrap();
+        let response = self.0.call1(&this, &request).map_err(|_| Error::msg("JavaScript fetcher throws an error."))?;
```
Comment (banaag) [inline sxg_rs/src/fetcher/js_fetcher.rs:45]: Could you include the error detail in the message?

File: sxg_rs/src/signature/js_signer.rs [hunk 3]
```diff
@@ -56,13 +57,13 @@ impl JsSigner {
 
 #[async_trait(?Send)]
 impl Signer for JsSigner {
-    async fn sign(&self, message: &[u8]) -> Result<Vec<u8>,String> {
+    async fn sign(&self, message: &[u8]) -> Result<Vec<u8>> {
         let a = Uint8Array::new_with_length(message.len() as u32);
         a.copy_from(&message);
         let this = JsValue::null();
-        let sig = self.js_function.call1(&this, &a).map_err(|_| "Bad call")?;
+        let sig = self.js_function.call1(&this, &a).map_err(|_| Error::msg("JavaScript signer throws an error."))?;
```
Comment (banaag) [inline sxg_rs/src/signature/js_signer.rs:64]: Are more error details possible?

File: sxg_rs/src/signature/js_signer.rs [hunk 4]
```diff
@@ -56,13 +57,13 @@ impl JsSigner {
...
         let sig = wasm_bindgen_futures::JsFuture::from(js_sys::Promise::from(sig));
-        let sig = sig.await.map_err(|_| "Bad async result")?;
+        let sig = sig.await.map_err(|_| Error::msg("JavaScript signer throws an error asynchronously."))?;
```
Comment (banaag) [inline sxg_rs/src/signature/js_signer.rs:66]: Same here, more error details would be better.

File: sxg_rs/src/signature/js_signer.rs [hunk 5]
```diff
@@ -73,9 +74,9 @@ impl Signer for JsSigner {
     }
 }
 
-fn raw_sig_to_asn1(raw: Vec<u8>) -> Result<Vec<u8>, String> {
+fn raw_sig_to_asn1(raw: Vec<u8>) -> Result<Vec<u8>> {
     if raw.len() != 64 {
-        return Err(format!("Expecting signature length to be 64, found {}.", raw.len()));
+        return Err(Error::msg(format!("Expecting signature length to be 64, found {}.", raw.len())));
```
Comment (banaag) [inline sxg_rs/src/signature/js_signer.rs:79]: Should the 64 be declared a constant somewhere? and then passed into this error message construct?

File: sxg_rs/src/sxg.rs [hunk 6]
```diff
@@ -12,14 +12,16 @@
...
 
+use anyhow::{Error, Result};
+
 // https://wicg.github.io/webpackage/draft-yasskin-httpbis-origin-signed-exchanges-impl.html#application-signed-exchange
-pub fn build(fallback_url: &str, signature: &[u8], signed_headers: &[u8], payload_body: &[u8]) -> Result<Vec<u8>, String> {
+pub fn build(fallback_url: &str, signature: &[u8], signed_headers: &[u8], payload_body: &[u8]) -> Result<Vec<u8>> {
     // https://wicg.github.io/webpackage/draft-yasskin-http-origin-signed-responses.html#name-application-signed-exchange
     if signature.len() > 16384 {
-        return Err("sigLength is larger than 16384".into())
+        return Err(Error::msg("sigLength is larger than 16384"));
```
Comment (banaag) [inline sxg_rs/src/sxg.rs:21]: These magic numbers (16384 and 524288) should be declared as constants.

File: sxg_rs/src/fetcher/js_fetcher.rs (hunk 1 above)
Comment (antiphoton) [reply sxg_rs/src/fetcher/js_fetcher.rs]: Done

File: sxg_rs/src/signature/js_signer.rs (hunk 3 above)
Comment (antiphoton) [reply sxg_rs/src/signature/js_signer.rs]: Done.

File: sxg_rs/src/signature/js_signer.rs (hunk 4 above)
Comment (antiphoton) [reply sxg_rs/src/signature/js_signer.rs]: Done.

File: sxg_rs/src/signature/js_signer.rs (hunk 5 above)
Comment (antiphoton) [reply sxg_rs/src/signature/js_signer.rs]: Done.

File: sxg_rs/src/sxg.rs (hunk 6 above)
Comment (antiphoton) [reply sxg_rs/src/sxg.rs]: Done.

**Review by banaag — APPROVED** [review]

**Pull request merged**